napari/components/viewer_model.py

```diff
@@ -354,16 +354,57 @@ def _on_active_layer(self, event):
     def rounded_division(min_val, max_val, precision):
         return int(((min_val + max_val) / 2) / precision) * precision
 
-    def _on_layers_change(self):
+    def _on_layers_change(self, *, add_layer=False):
+        new_point = None
         if len(self.layers) == 0:
             self.dims.ndim = 2
             self.dims.reset()
         else:
             ranges = self.layers._ranges
+            # If the point of an axis is outside of the upper current range,
+            # set it to the layers range max value. dims.set_current_step
+            # accounts for the clamping of the lower limit on the set_point call.
+            # New dimensions are prepended, hence we need to go through the list
+            # from the back to the beginning and invert the order in the end.
+            # To have the correct slices present.
+            prev_point = [
+                range_val[1] - 1 if point_val >= range_val[1] else point_val
+                for point_val, range_val in zip(
+                    reversed(self.dims.point), reversed(ranges)
+                )
+            ][::-1]
             ndim = len(ranges)
             self.dims.ndim = ndim
             self.dims.set_range(range(ndim), ranges)
 
+            # This logic block defines the behavior when layers are added or
+            # removed. Its goal is it to keep the currently selected view
+            # selected by the user.
+            if add_layer and len(self.layers) == 1:
+                # Special case going from 0 to 1 layer leads to the central
+                # slice being selected.
+                # Move the logic from _on_add_layer here to avoid double set
+                # and double update of the layers.
+                new_point = [
+                    self.rounded_division(*_range) for _range in ranges
+                ]
+            elif ndim < len(prev_point):
+                # If the dimension has been reduced remove dimensions accordingly.
+                # Since new dimensions are prepended we need to take the last points.
+                new_point = prev_point[-ndim:]
```

Comment:
I would have expected `test_dimension_change_is_visible_in_viewport` to cover this, but if I set a breakpoint here and debug that test, I do not hit this line, so it looks like codecov is correct. Would be good to cover this new code by modifying the code or adding a test.

Reply:
This is fixed by removing the custom logic of prev_point and let the adjustment happen by the `set_current_step` function.